proposals/subgroups.md

```diff
@@ -0,0 +1,227 @@
+
+## Built-in Values
+
+| Built-in | Type | Direction | Description |
+| --- | --- | --- | --- |
+| `subgroup_size` | u32 | Input | The size of the current subgroup |
+| `subgroup_invocation_id` | u32 | Input | The index of the invocation in the current subgroup |
+
+Note: HLSL does not expose a subgroup_id or num_subgroups equivalent.
+
+## Built-in Functions
+
+All built-in function can only be used in `compute` or `fragment` shader stages.
+Using f16 as a parameter in any of these functions requires `subgroups-f16` to be enabled.
+
+| Function | Preconditions | Description |
+| --- | --- | --- |
+| `fn subgroupElect() -> bool` | | Returns true if this invocation has the lowest subgroup_invocation_id among active invocations in the subgroup |
+| `fn subgroupAll(e : bool) -> bool` | | Returns true if `e` is true for all active invocations in the subgroup |
+| `fn subgroupAny(e : bool) -> bool` | | Returns true if `e` is true for any active invocation in the subgroup |
+| `fn subgroupBroadcast(e : T, id : I) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types<br>`I` must be i32 or u32 | Broadcasts `e` from subgroup_invocation_id `id` to all active invocations. `id` must be dynamically uniform<sup>1</sup> |
+| `fn subgroupBroadcastFirst(e : T) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Broadcasts `e` from the Broadcasts `e` from the active invocation with the lowest subgroup_invocation_id in the subgroup  to all other active invocations |
+| `fn subgroupBallot(pred : bool) -> vec4<u32>` | | Returns a set of bitfields where the bit corresponding to subgroup_invocation_id is 1 if `pred` is true for that active invocation and 0 otherwise. |
+| `fn subgroupShuffle(v : T, id : I) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types<br>`I` must be u32 or i32 | Returns `v` from the active invocation whose subgroup_invocation_id matches `id` |
+| `fn subgroupShuffleXor(v : T, mask : u32) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Returns `v` from the active invocation whose subgroup_invocation_id matches `subgroup_invocation_id ^ mask` |
+| `fn subgroupShuffleUp(v : T, delta : u32) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Returns `v` from the active invocation whose subgroup_invocation_id matches `subgroup_invocation_id - delta` |
+| `fn subgroupShuffleDown(v : T, delta : u32) -> T` | `T` must be u32, i32, f32, f16 or a vector of those types | Returns `v` from the active invocation whose subgroup_invocation_id matches `subgroup_invocation_id + delta` |
+| `fn subgroupSum(e : T) -> T` | `T` must be u32, i32, f32, or a vector of those types | Reduction<br>Adds `e` among all active invocations and returns that result |
+| `fn subgroupInclusiveSum(e : T) -> T)` | `T` must be u32, i32, f32, f16 or a vector of those types | Inclusive scan<br>Returns the sum of `e` for all active invocations with subgroup_invocation_id less than or equal to this invocation |
```

Comment:
<details><summary>subgroupInclusiveSum naming bikeshedding</summary>

It's not obvious from the name this behaves differently from `subgroupSum` in that it returns a unique value per invocation rather than a subgroup-uniform value. Not having prior knowledge I'd think a standard sum operation was already "inclusive".

I think `subgroupPrefixInclusiveSum` or `subgroupInclusivePrefixSum` would be much more obvious at the cost of verbosity. Alternatively, `subgroupPrefixSum` still hints that something more complex than an ordinary sum is going on to an unfamiliar reader, but for experienced users leaves room for confusion as to the in/exclusivity. It doesn't help that DX12 `WavePrefixSum()` is [exclusive](https://github.com/microsoft/DirectXShaderCompiler/blob/8147212ee532845e39db0f9b0a2dd5fe887b814f/docs/SPIR-V.rst?plain=1#L3857)

</details>

Reply:
It is meant to be exclusive.

Reply:
Bikeshedding thoughts.

Comparing prior art:
* "Prefix sum" is what I've known the general algorithm as, and likely the first entry for most folks.  https://en.wikipedia.org/wiki/Prefix_sum 
* HLSL and MSL use "prefix" in the names.  GLSL does not.
* HLSL and MSL use "sum" and "product".  GLSL uses "Add" and "Mul".   The GLSL way generalizes to logical operations like And, Or, Max, Min (or those are simply ambiguoous!  Hmmm.)

Once I see many operations in context like in [this Vulkan subgroups tutorial](https://www.khronos.org/blog/vulkan-subgroup-tutorial#:~:text=T%20subgroupInclusiveAdd(T%20value)), lacking "prefix" doesn't seem so bad because it's implied by `Inclusive` or `Exclusive`.